Extend debugger interface

jdtzmn opened the pull request:
## Summary

This draft PR includes an extended debugger interface which extends the existing `logic.DebuggerHooks` interface with additional events which will become useful within the simulator.

The PR currently points to the wrong base to make it easier to see what files have been changed since then while gathering initial reviews.

## Test Plan

- [x] Additional tests in `debugger_test.go` for the new hooks
- [x] Test `eval.TransactionGroupWithDebugger()` within `eval_test.go`

jasonpaulos commented on daemon/algod/api/server/v2/dryrun.go:
```diff
@@ -182,23 +182,23 @@ func (ddr *dryrunDebugReceiver) stateToState(state *logic.DebugState) generated.
-func (ddr *dryrunDebugReceiver) Complete(state *logic.DebugState) error {
-	return ddr.Update(state)
+func (ddr *dryrunDebugReceiver) AfterAppEval(state *logic.DebugState) error {
+	return ddr.AfterAppEval(state)
```
This does look to be a recursive call. Is that a mistake?

jdtzmn replied:
Yes, it most definitely is. Only a little concerning that the dryrun tests didn't catch it.

jasonpaulos commented on daemon/algod/api/server/v2/dryrun.go:
```diff
@@ -182,23 +182,23 @@ func (ddr *dryrunDebugReceiver) stateToState(state *logic.DebugState) generated.
 	return nil
 }
 
 // Update is fired on every step (DebuggerHook interface)
```
nit: function names in comments are outdated.

Also, the behavior of this `dryrunDebugReceiver` will now change, since `BeforeTealOp` will be invoked for inner txn ops as well? And previously it was not?

jdtzmn replied:
It previously was as well. For the most part, `BeforeTealOp` is a renaming of `Update`. See [these lines](https://github.com/jdtzmn/go-algorand/pull/4/files#diff-7efbb393ce12c86b88888ff627dd385ba50ec9937b26082263ef580068080d27L722-R722) in `logic/eval.go`.

jasonpaulos replied:
Previously `NewInnerEvalParams` did not propagate the `Debugger` property from the parent `EvalParams`, and in this PR it does. So I think the current dryrun would see a different behavior here.

jdtzmn replied:
That's a great point. I hadn't thought of that. I'm assuming this new behavior is something we don't want. A simple solution would be to have a flag which disables the teal op logic when within an inner transaction. Do you think that would be sufficient?

jasonpaulos replied:
Yep, that sounds like a good plan

jasonpaulos commented on data/transactions/logic/eval.go:
```diff
@@ -680,11 +681,9 @@ func eval(program []byte, cx *EvalContext) (pass bool, err error) {
-			if err == nil {
-				err = errDbg
-			}
+		errDbg := callAfterAppEvalHookIfItExists(cx.Debugger, cx.refreshDebugState(err))
```
One downside of this style of method invocation: `cx.refreshDebugState` will be called even if `cx.Debugger` is nil, or if `cx.Debugger` does not implement this specific method. `cx.refreshDebugState` copies the entire stack and scratch space, so this is potentially an expensive operation to invoke frequently.

jasonpaulos commented on data/transactions/logic/debugger.go:
```diff
@@ -34,14 +34,157 @@ import (
+
+type debuggerBeforeInnerTxnHook interface {
+	// BeforeInnerTxn is called before the inner transaction is executed
+	BeforeInnerTxn(ep *EvalParams) error
```
Shouldn't group index be passed here as well, otherwise there's no easy to way to tell which txn is being executed?

jdtzmn replied:
No because this hook is meant to indicate that an inner transaction is about to be executed (think something equivalent to a non-existent `BeforeTxnGroup`).

Instead, the `BeforeTxn` hook is _supposed to be_ re-used before each transaction submitted within an `itxn_submit`. However, it seems I lost track of that when implementing the hooks within the evaluation flow and so it's not the case currently. Once I fix that, you can use `BeforeTxn` to get each inner transaction's group index.

jasonpaulos replied:
Ok, that makes sense then. I'd suggest clarifying this in the docs so that it's clear for future readers too. And perhaps renaming to `BeforeInnerTxnGroup`?

jdtzmn replied:
Good idea! I'll rename it and update the docs.

jasonpaulos commented on data/transactions/logic/debugger.go:
```diff
@@ -34,14 +34,157 @@ import (
+	// ┌────────────────────────────────────────────────┐
+	// │ Transaction Evaluation                         │
+	// ├────────────────────────────────────────────────┤
+	// │ > BeforeTxn                                    │
```
Have you thought about if/how the interface can support LogicSig evaluation as well?

jdtzmn replied:
I had not. Glad someone was thinking about that 😅 . I've since added implementation details.

jasonpaulos reviewed (approved):
This looks good!

My one comment is that it would be nice to see a unit test or two for the changes in `data/transactions/verify/txn.go`, but that can be done in a different place before release

michaeldiamant committed:
Tests: Fix handling of empty local state in TestResourcesDataApp (#4699)

algoidurovic committed:
AVM: match, pushints, and pushbytess opcodes (#4645)

add match opcode along with assembler and eval unit tests
implementation of match, pushints, and pushbytess
add docs for pushints and pushbytess

cce committed:
network: align duplicateFilterCount to prevent panic on 32-bit platforms (#4702)

AlgoStephenAkiki committed:
Bug-fix: return v2 spec from swagger endpoint (#4697)

michaeldiamant committed:
AVM:  Add support for Boxes (#4149)

jasonpaulos committed:
Merge branch 'master' into feature/simulate-endpoint

jasonpaulos committed:
Fix generate config files

cce committed:
performance: don't create child cow in TestTransactionGroup (#4706)

michaeldiamant committed:
db: print dbutil.go error in stderr (#4712)

jasonpaulos committed:
Merge branch 'feature/simulate-endpoint' into extend-debugger-interface

iansuvak committed:
performance: agreement state serialization using msgp (#4644)

algorandskiy committed:
prefetcher: enable prefetcher for assets and apps (#4352)

jasonpaulos committed:
Reorder debugger hook functions

algorandskiy committed:
test: fix after boxes and prefetcher remerge (#4716)

tzaffi committed:
blank import needs commenting out for indexer's lint (#4720)

jannotti committed:
AVM: fix early eval exits for Debugger (#4719)


Co-authored-by: Michael Diamant <michaeldiamant@users.noreply.github.com>

onetechnical committed:
DevOps: Update docker container in build (#4723)

algorandskiy committed:
catchpoint: actualize limit constants and fix txtail table init (#4703)

barnjamin committed:
tools: allow specification of RewardPoolBalance in genesis (#4643)

jannotti committed:
Consensus: v36 upgrade (#4722)

v36 consensus upgrade

jasonpaulos committed:
Merge branch 'master' into feature/simulate-endpoint

jasonpaulos committed:
Merge branch 'feature/simulate-endpoint' into extend-debugger-interface

algolucky committed:
dispenser: add input validation (#4713)

Algo-devops-service committed:
Update the Version, BuildNumber, genesistimestamp.data

algojohnlee committed:
Merge pull request #4730 from Algo-devops-service/relbeta3.12.0

algolucky committed:
mergeback: feature/alphanet  (#4726)

cce committed:
add unconvert to warning linter (#4710)

algorandskiy committed:
db: add AtomicContext (#4724)

Co-authored-by: Tolik Zinovyev <tolik@algorand.com>

cce committed:
update vAlpha5 to point to v36 (#4729)

algorandskiy committed:
tests: make ledger tests ~20 sec faster (#4739)

almog-t committed:
catchpointdump: supports tar.gz files (#4743)

iansuvak committed:
perf: don't preallocate rarely used maps in MakeStateDelta (#4715)

winder committed:
tools: remove doberman and dependency (#4750)

algochoi committed:
algod: Migrate internal uses of v1 algod API to v2 (#4684)

michaeldiamant committed:
test:  add test defending against regression in testnet genesis generation (#4741)

algochoi committed:
tests: Fix account tests for calling pending txn endpoint and box tests (#4751)

winder committed:
goal: Add URL to goal asset info. (#4731)

icorderi committed:
performance: turn cache misses during assembly into cache hits during eval (#4617)

jdtzmn committed:
Testing: Lint for unparallelized tests without explicit reason (#4746)

algorandskiy committed:
daemon: add pool error message (#4755)

Txpool might store failed transactions for quite a long time on its StatusCache.
When someone requests a pending transaction and it gets retrieved from the cache,
there is no indication it actually failed, and a user might have an impression the txn is fine but stuck.

Co-authored-by: Jason Paulos <jasonpaulos@users.noreply.github.com>

algochoi committed:
Tests: Fix types in assertions and change to using libgoal client when getting `PendingTransaction` information (#4754)

algochoi committed:
goal: Change asset commands to use AccountAssetInformation (#4758)

Eric-Warehime committed:
API: Update oapi-codegen version to v1.12.0 of our fork (#4707)

Updates the version of oapi-codegen which is installed via
the buildtools to our v1.12.0 release. Also changes the
Makefile targets and adds config files which are the preferred
way of configuration instead of the old CLI flags.

The generated types definitions have also been updated using
the oapi-codegen commit which ensures all required Enums are generated
for enums embedded in response types.

algochoi committed:
Tests: Move box test artifacts to a temp directory (#4771)

jasonpaulos committed:
WIP

algochoi committed:
algod: Delete v1 usage from client and remove `V2` prefix from APIs (#4767)

algolucky committed:
netgoal: start with 20GB RootStorage (#4779)

algonoah committed:
Update SECURITY.md (#4782)

onetechnical committed:
Bump buildnumber.dat

cce committed:
check for large Topics length

onetechnical committed:
Merge pull request #5 from algorand/relstable3.11.3

go-algorand v3.11.3-stable

Eric-Warehime committed:
Fix resolver test endpoint (#4784)

Eric-Warehime committed:
algod: Consolidate models (#4714)

michaeldiamant committed:
Notify when GHA workflow fails on master (#4791)

michaeldiamant committed:
Notify about CircleCI build failures for master and rel branches only (#4789)

algochoi committed:
Tests: Fix txid-string comparison (#4792)

Algo-devops-service committed:
Merge remote-tracking branch 'origin/rel/stable' into relstable3.11.3-remerge

algojohnlee committed:
Merge pull request #4811 from Algo-devops-service/relstable3.11.3-remerge

mergeback: Relstable3.11.3 remerge

cce committed:
telemetry: add TCP RTT info collection (#4745)

algorandskiy committed:
txHandler: add more metric (#4786)

* add txHandler backlog error reason counters
* add txHandler backlog size gauge
* add txsync counters
* add TX, AV, PP, MI counters to
  DisconnectPeerEventDetails and PeerConnectionDetails

Co-authored-by: chris erway <chris.erway@algorand.com>

iansuvak committed:
metrics: split out /Transaction/AssembleBlock metrics (#4795)

brianolson committed:
test: large generated accountdb cluster test (#4772)

cce committed:
cow: always process KvMods to add OldData to mods when building StateDeltas (#4804)

algonautshant committed:
txHandler: service does not restart after node catching up (#4809)

jannotti committed:
Catchpoints: Add the kv hashes into the trie (#4812)

* Add the kv hashes into the trie

* Periodically commit, and report proper stats

algonautshant committed:
txSyncer: service does not start after node catching up (#4768)

michaeldiamant committed:
CI:  Fix GHA failure notification trigger condition (#4810)

michaeldiamant committed:
Catchpoints:  Enrich catchpoint generation and status with KV metadata (#4808)

michaeldiamant committed:
Squelch macOS accept incoming network connections notifications (#4814)

cce committed:
add TransactionGroupTxSyncHandled (#4815)

cce committed:
tools: update catchpointdump 'database' and 'database check' commands to handle staging tables and KVs (#4802)

* update catchpointdump 'database' and 'database check' commands to handle staging tables and KVs

* address CR comments

cce committed:
accountdb: add migration 8 for resetting accounthashes on betanet (#4822)

Algo-devops-service committed:
Merge branch 'master' into relbeta3.12.1

Algo-devops-service committed:
Update the Version, BuildNumber, genesistimestamp.data

icorderi committed:
ledger: move pieces of accountdb.go into a storage package  (#4776)

algorandskiy committed:
test: disable TestWebsocketNetworkTelemetryTCP on CI (#4827)

algorandskiy committed:
test: disable TestWebsocketNetworkTelemetryTCP on CI (#4827)

algojohnlee committed:
Merge pull request #4823 from Algo-devops-service/relbeta3.12.1

onetechnical committed:
Bump buildnumber.dat

algojohnlee committed:
Merge pull request #4828 from onetechnical/relbeta3.12.2

icorderi committed:
ledger: move catchpoint sql code into storage package (#4813)

michaeldiamant committed:
ledger: add context to block mismatch error in VerifyCatchpoint (#4821)

iansuvak committed:
performance: implement sync.Pool for cow.child() (#4766)

icorderi committed:
ledger: move non-lookup methods into a storage package (#4830)

algojohnlee committed:
Merge pull request #4837 from Algo-devops-service/relstable3.12.2

algorandskiy committed:
metrics: count large txn groups and msigs (#4833)

Eric-Warehime committed:
node: fix disabled node tests (#4824)

icorderi committed:
ledger: move catchpoint related methods into a storage package (#4835)

ahangsu committed:
Bug-Fix: upgrade avm-abi (#4838)

michaeldiamant committed:
Tests: Extend TestCatchpointAfterTxns to catch catchpoint write corruption (#4818)

michaeldiamant committed:
Catchpoint: Fix broken merge between #4818 and #4835 (#4840)

icorderi committed:
ledger: move catchpoint staging into a storage package (#4836)

Algo-devops-service committed:
Bump Version, Remove buildnumber.dat and genesistimestamp.dat files.

Algo-devops-service committed:
Merge remote-tracking branch 'origin/rel/stable' into relstable3.12.2-remerge

algojohnlee committed:
Merge pull request #4843 from Algo-devops-service/relstable3.12.2-remerge

CICD: go-algorand relstable3.12.2-remerge mergeback

algonautshant committed:
txHandler: upgraded benchmarks (#4834)

* Add setting incoming transaction rate
* Add backlog worker with transaction dropping
* Count dropped transactions

algolucky committed:
updater: use more-recent installer package (#4734)

icorderi committed:
ledger: move blockdb into a storage package (#4841)

Eric-Warehime committed:
Ledger Deltas: AccountDelta retrieval through ledger/catchup service (#4658)

Co-authored-by: shiqizng <80276844+shiqizng@users.noreply.github.com>
Co-authored-by: michaeldiamant <michaeldiamant@users.noreply.github.com>
Co-authored-by: Will Winder <wwinder.unh@gmail.com>

jannotti committed:
CI: Fix some details around the hashKind -> HashKind change (#4865)

* Fix some details around the hashKind -> HashKind change

* CI should do full `make generate`

AlgoStephenAkiki committed:
goal: account deletepartkey subcomand (#4788)

icorderi committed:
ledger: move migrations into a storage package (#4846)

icorderi committed:
ledger: move blockdb tests into storage package (#4860)

jasonpaulos committed:
Merge branch 'master' into feature/simulate-endpoint

jasonpaulos committed:
Clarify group transactions are also acceptable

jasonpaulos committed:
Try to appease the new parallel linter

algochoi committed:
algod: Sunset v1 handlers (#4847)

jasonpaulos committed:
Goal: Test ABI reference arguments during creation (#4872)

algorandskiy committed:
txHandler: early duplicates detection (#4806)

Add tx message duplicates detection in txHandler:
  * Network raw messages with blake2b + salt auto-rotating hash map
  * Canonical reencoded tx groups with sha512_256 hash map
Configurable with TxIncomingFilteringFlags

Co-authored-by: chris erway <chris.erway@algorand.com>
Co-authored-by: algonautshant <shant@algorand.com>

icorderi committed:
ledger: new db methods to abstract queries needed by the deltas (#4864)

winder committed:
DevOps: General purpose docker container. (#4816)

algorandskiy committed:
txHandler: metrics for txpool.Remember (#4839)

algorandskiy committed:
txHandler: disable canonical hashmap by default (#4877)

algochoi committed:
algod: Delete deprecated cost field in dyrun response and mapping field in sourcemap (#4875)

algorandskiy committed:
metrics: fix TestGetValue (#4879)

cce committed:
metrics: add counters for checkAlreadyCommitted (#4881)

* breakdown of eval vs block for lease/txID errors

Eric-Warehime committed:
tests: skip flaky node tests (#4862)

Eric-Warehime committed:
algod: Remove unused generated code (#4880)

cce committed:
prefetcher: preallocate aligned groupTasks in groupsReady (#4876)

michaeldiamant committed:
catchpoint:  promote MT-related log warnings to error (#4882)

Algo-devops-service committed:
Update the Version, BuildNumber, genesistimestamp.data

algojohnlee committed:
Merge pull request #4885 from Algo-devops-service/relbeta3.13.0

jasonpaulos committed:
Classify simulation endpoint as experimental

jasonpaulos committed:
Merge branch 'master' into feature/simulate-endpoint

algorandskiy committed:
testing: defi txn group benchmark (#4886)

Co-authored-by: John Jannotti <jannotti@gmail.com>

algojohnlee committed:
Merge pull request #4890 from Algo-devops-service/relstable3.13.0

jasonpaulos committed:
Rename EnableTransactionSimulator to EnableExperimentalAPI

jasonpaulos committed:
Update config.json.example

jasonpaulos committed:
Merge branch 'feature/simulate-endpoint' into extend-debugger-interface

jasonpaulos committed:
File renames

barnjamin committed:
Langspec: adding version to langspec (#4883)

algorandskiy committed:
teal: optimize CheckSignature (#4893)

cce committed:
metrics: split alive check counter between early/dead (#4898)

winder committed:
rest api: make the data endpoints public (#4900)

Eric-Warehime committed:
docs: update API docs, remove unused codegen configs (#4896)

algorandskiy committed:
metrics: update some new metric names (#4899)

Algo-devops-service committed:
Merge branch 'master' into relbeta3.13.1

Algo-devops-service committed:
Update the Version, BuildNumber, genesistimestamp.data

barnjamin committed:
Enhancement: Add missing ArgEnum fields, tweaks to immediate note (#4903)

* Adding argenum fields, tweaks to immediate note

* Fix fmt

* restore big-endian note

* adding reminder to possibly modify argEnums method with some automatic thing later

jasonpaulos committed:
Rename DebuggerHook to EvalTracer and remove error return values from hooks

jasonpaulos committed:
Merge branch 'master' into feature/simulate-endpoint

jasonpaulos committed:
Make experimental routes public

jasonpaulos committed:
Add to API docs

jasonpaulos committed:
Missing change to README

jasonpaulos committed:
Linter bug

jasonpaulos committed:
Merge branch 'feature/simulate-endpoint' into extend-debugger-interface

jasonpaulos committed:
debuggerEvalTracerAdaptor embeds NullEvalTracer

algorandskiy committed:
tests: fix flaky TestTxHandlerRememberReportErrorsWithTxPool (#4905)

jasonpaulos committed:
Rename EvalTracer methods

jasonpaulos committed:
Get rid of legacy names

jasonpaulos committed:
Better use of txntest package

jasonpaulos committed:
Let's see if the export linter yells at this

jasonpaulos committed:
Rename DebuggerHook to Debugger

icorderi committed:
ledger: move iterators into a store package (#4874)

* rename remaining accountdb to acctdeltas

jasonpaulos committed:
Ensure tracer implements don't drift from interface

jasonpaulos committed:
Add back export comments

jasonpaulos committed:
Improve EvalTracer method docs

algojohnlee committed:
Merge pull request #4904 from Algo-devops-service/relbeta3.13.1

Algo-devops-service committed:
Merge branch 'rel/beta' into relstable3.13.1

algojack committed:
DevOps: cleanup old artifact on self-hosted runners #4909

jasonpaulos committed:
Add experimental warning to swagger file

jasonpaulos committed:
Merge branch 'master' into feature/simulate-endpoint

jasonpaulos committed:
Merge branch 'feature/simulate-endpoint' into extend-debugger-interface

algojohnlee committed:
Merge pull request #4915 from Algo-devops-service/relstable3.13.1

icorderi committed:
ledger: add kvIter to storage package (#4892)

algorandskiy committed:
testing: make MakeAccountData as testing only method (#4856)

algorandskiy committed:
rest api: update WaitForBlock summary (#4910)

algonautshant committed:
txHandler: msig and lsig benchmarks (#4908)

Co-authored-by: Pavel Zbitskiy <65323360+algorandskiy@users.noreply.github.com>

algonautshant committed:
tests: fix dependency on global counters (#4921)

algorandskiy committed:
bench: fix benchmarking after remerges (#4923)

jdtzmn committed:
Algod: Add experimental endpoint for simulating transactions against a real block evaluator (#4436)

Co-authored-by: Michael Diamant <michaeldiamant@users.noreply.github.com>
Co-authored-by: Jason Paulos <jasonpaulos@users.noreply.github.com>

Algo-devops-service committed:
Bump Version, Remove buildnumber.dat and genesistimestamp.dat files.

Algo-devops-service committed:
Merge remote-tracking branch 'origin/rel/stable' into relstable3.13.1-remerge

algojohnlee committed:
Merge pull request #4925 from Algo-devops-service/relstable3.13.1-remerge

algonautshant committed:
bench: use more common multisig sizes (#4926)

almog-t committed:
ledger: fix redundant unlock potentially causing double unlock (#4928)

cce committed:
telemetry: don't report let asyncTelemetryHook wrap and log its own errors (#4932)

cce committed:
telemetry: don't report let asyncTelemetryHook wrap and log its own errors (#4932)

algorandskiy committed:
log: stderr offending large log messages (#4930)

algorandskiy committed:
log: stderr offending large log messages (#4930)

cce committed:
count telemetry server errors (#4933)

cce committed:
count telemetry server errors (#4933)

algobarb committed:
Bump buildnumber.dat

jdtzmn committed:
Tests: Fix `t.Parallel()` errors in `data/transactions/logic` package (#4931)

algojohnlee committed:
Merge pull request #4936 from algobarb/relbeta3.13.2

michaeldiamant committed:
Tests: Enable more parallel tests in data/transactions/logic/evalAppTxn_test.go (#4937)

Algo-devops-service committed:
Merge branch 'rel/beta' into relstable3.13.2

algojohnlee committed:
Merge pull request #4938 from Algo-devops-service/relstable3.13.2

michaeldiamant committed:
tests: switch e2e_basic_start_stop from devnet to private network (#4942)

* Disable TestResolver in CircleCI due to DNS resolution issue
* Switch e2e_basic_start_stop from devnet to private network

michaeldiamant committed:
tests:  remove unused Python cryptography package (#4939)

algobarb committed:
remove buildnumber.dat

algojohnlee committed:
Merge pull request #4945 from algobarb/algobarb/relstable3.13.2-remerge

AlgoAxel committed:
txHandler: Random Early Detection for backlog queue (#4797)

* Elastic Rate Limiter and Congestion Manager to enable RED in txHandler.
* This RED model implements fairness, in which the chance a client's request is dropped
   is proportional to their individual arrival rate, vs a per-client service rate.
* Backlog size is increased by ~30% to accommodate some guaranteed throughput at current TPS target rate.

cce committed:
telemetry: don't report ES client errors (#4940)

algonautshant committed:
txHandler: batch transactions to the verifier (#4621)

* Batching incoming transaction for crypto verification
* Streaming capabilities of transactions to the verifier

id-ms committed:
update copyright to 2023 (#4957)

michaeldiamant committed:
tests: split TestResolver to enable more tests on CircleCI (#4944)

michaeldiamant committed:
Telemetry: Remove unused telemetry StartOperation (#4935)

brianolson committed:
remove unused node PoolStats() (#4956)

mciccarello committed:
goal: Added consensus upgrade fields to node status api & goal output (#4800)

brianolson committed:
plot data in algod_go_memory_classes_heap_objects_bytes (#4968)

almog-t committed:
ledger: bulletin leaking waiters fix (#4962)

brianolson committed:
tools: nodecfg without dns (#4967)

Eric-Warehime committed:
node: add logging thread to node monitoring wg (#4934)

michaeldiamant committed:
tests: re-enable TestResolverWithCloudflareDNSResolution in CircleCI (#4965)

algolucky committed:
util/s3: enable more credential providers (#4929)

algolucky committed:
CICD: add container build (#4927)

jannotti committed:
AVM: Catch any panic in edcsa verifying (#4368)

algonautshant committed:
tests: fix flaky TestStreamVerifierCtxCancelPoolQueue (#4980)

jdtzmn committed:
Tests: Fix `t.Parallel()` errors in `data` package (#4981)

cce committed:
metrics: use uint64 for Counter and Gauge types (#4911)

jannotti committed:
Revert "AVM: Catch any panic in edcsa verifying (#4368)" (#4985)

This reverts commit 19fa6c2471181d494245b0a36d7d8d57ca18ddab.

AlgoAxel committed:
tests: fix unstable TestREDCongestionManagerShouldntDrop (#4986)

algolucky committed:
fix(scripts): use -z in conditional (#4987)

cce committed:
config: add HeartbeatUpdateInterval (#4832)

onetechnical committed:
Bugfix: update.sh breaks fetching latest due to new anon cred line (#4994)

gmalouf committed:
Revert "Tests: Fix `t.Parallel()` errors in `data` package (#4981)" (#4995)

This reverts commit c41cd69afbd33ad543a70fd319de77bffc0aa4fd.

gmalouf committed:
Retain improvements made in #4981 PR that had been reverted due to instability. the changes here should net improve test reliability and rationale for why several are not marked parallel. (#4996)

michaeldiamant committed:
CI: Split paralleltest lint excludes to minimize merge conflicts (#5002)

jasonpaulos committed:
Merge branch 'master' into extend-debugger-interface

jasonpaulos committed:
Fix test issue

jasonpaulos committed:
Update license headers

jannotti committed:
AVM: Clearer cost benchmarks, and a nice optimization of b== and b< (#5010)

jasonpaulos committed:
minor updates from CR

jasonpaulos committed:
Use assembler instead of hardcoding program bytes

tzaffi committed:
goal: TEAL assembler error UX improvement (#4951)

Changed assembler functionality so that logic.assemble errors when receiving empty input

jasonpaulos committed:
Make tracer a field in BlockEvaluator and generalize txn group hooks

jasonpaulos committed:
Don't add tracer to GroupContext

jasonpaulos committed:
Fix incorrect txn depth logic in debuggerEvalTracerAdaptor

algolucky committed:
container: update documentation and remove unused tools (#4982)

algorandskiy committed:
tools: box key utility (#5026)

algojack committed:
Bump version in buildnumber.dat

AlgoStephenAkiki committed:
avm-abi: Update functions migrated to avm-abi library (#4979)

algorandskiy committed:
Empty box hotfix for relstable (#16)

onetechnical committed:
Merge pull request #17 from algorand/relstable3.13.3

go-algorand v3.13.3-stable

algorandskiy committed:
fix build after #4979 (#5032)

winder committed:
Remove unused NetworkFetcher code. (#5028)

brianolson committed:
ledger: rearrange blockqueue start/stop (#4964)

algolucky committed:
util/s3: remove extra println (#5024)

cce committed:
debug: unmarshal and print transaction distribution (#5008)

winder committed:
Remove accidental devtesting output. (#5033)

winder committed:
Remove topAccountListener (#5027)

brianolson committed:
tools: pingpong total latency (#4757)

cce committed:
config: add CadaverDirectory and prevent algod from starting if cadaver.init() fails (#4740)

onetechnical committed:
Merge branch 'rel/stable' into relstable-3-13-3-remerge

algojohnlee committed:
Merge pull request #5036 from algorand/onetechnical/relstable-3-13-3-remerge

algorandskiy committed:
network: increase max incoming connections limit to 2400 (#5018)

* Decrease connections per ip addr limit to 15

algojack committed:
CI: Fix PartitionTest Github Action (#5006)

* Updating action versions
* Removing inlining and optimizations from building partitiontest-linter
* Adding flags to golangci-lint
* Syncing typeparams dependency version
* Changing tools package version
* running go mod tidy and adding compat 1.17
* Chris's fix for linter
* Filtering on added only, upgrading reviewdog, diplaying errors in log

michaeldiamant committed:
Tests: Fix t.Parallel() errors in shared package (#4989)

* Fix t.Parallel() errors in shared package
* Correct .golangci.yml paths

jasonpaulos committed:
Merge branch 'master' into extend-debugger-interface
